pkg/dogstatsd/server.go

```diff
@@ -547,11 +569,49 @@ func (s *Server) errLog(format string, params ...interface{}) {
 	}
 }
 
-func (s *Server) parseMetricMessage(metricSamples []metrics.MetricSample, parser *parser, message []byte, origin string) ([]metrics.MetricSample, error) {
+// createOriginTagMaps  keeps these in a cache to avoid a lot of heap escape
+// that we can't avoid in this hot path
+func (s *Server) createOriginTagMaps(origin string) cachedTagsOriginMap {
+	okMap := map[string]string{"message_type": "metrics", "state": "ok"}
+	errorMap := map[string]string{"message_type": "metrics", "state": "error"}
+	okMap["origin"] = origin
+	errorMap["origin"] = origin
+	maps := cachedTagsOriginMap{
+		origin: origin,
+		ok:     okMap,
+		err:    errorMap,
+	}
+
+	s.cachedTlmOriginIds[origin] = maps
+	s.cachedOrder = append(s.cachedOrder, maps)
+
+	if len(s.cachedOrder) > maxOriginTagsCached {
+		// remove the oldest one from the cache
+		pop := s.cachedOrder[0]
+		delete(s.cachedTlmOriginIds, pop.origin)
+		s.cachedOrder = s.cachedOrder[1:]
+		// remove it from the telemetry metrics as well
+		tlmProcessed.DeleteWithTags(pop.ok)
+		tlmProcessed.DeleteWithTags(pop.err)
+	}
+
+	return maps
+}
+
+func (s *Server) parseMetricMessage(metricSamples []metrics.MetricSample, parser *parser, message []byte, origin string, telemetry bool) ([]metrics.MetricSample, error) {
 	sample, err := parser.parseMetricSample(message)
 	if err != nil {
 		dogstatsdMetricParseErrors.Add(1)
-		tlmProcessed.IncWithTags(tlmProcessedErrorTags)
+		if origin != "" && telemetry {
+			var maps cachedTagsOriginMap // errorMap and okMap for this origin
+			var exists bool
+			if maps, exists = s.cachedTlmOriginIds[origin]; !exists {
+				maps = s.createOriginTagMaps(origin)
+			}
+			tlmProcessed.IncWithTags(maps.err)
+		} else {
+			tlmProcessed.IncWithTags(tlmProcessedErrorTags)
+		}
```

Comment:
Could we have a generic function to avoir duplication with line 649 ?

Reply:
Do you feel that it would really make it clearer? Since both are not using the same objects (L605 is using error and L649 is using ok state variables), passing all these objects to that function I would create would not look good IMO. Please let me know if you really prefer that, I don't have a strong opinion on that except that I would have to make sure it's not creating any heap escape 😄

Reply:
I don't have a strong opinion either.